plugins/connection/redirection_utility.c

```diff
@@ -0,0 +1,555 @@
+				NULL))
+			{
+				mutex_unlock(redirect_mutex);
+				return 0;
+			}
+
+			ma_strmake(host_str, ret_entry->host, strlen(ret_entry->host));
+			ma_strmake(user_str, ret_entry->user, strlen(ret_entry->user));
+			(*info)->host = host_str;
+			(*info)->user = user_str;
+			(*info)->port = ret_entry->port;
+		}
+	}
+
+	my_bool retval = (ret_entry != NULL);
+	mutex_unlock(redirect_mutex);
+
+	return retval;
+}
+
+void add_redirection_entry(char* key, char* host, char* user, unsigned int port)
+{
+	struct Redirection_Entry* new_entry = NULL;
+	struct Redirection_Entry* popingEntry = NULL;
+	char* _key;
+	char* _host;
+	char* _user;
+	size_t key_len = 0;
+	size_t host_len = 0;
+	size_t user_len = 0;
+
+	if (key == NULL || host == NULL || user == NULL)
+	{
+		return;
+	}
+
+	my_bool copy_on_found = 0;
+	if (get_redirection_info(key, NULL, copy_on_found))
+	{
+		return;
+	}
+
+	key_len = strlen(key);
+	host_len = strlen(host);
+	user_len = strlen(user);
+
+	if (!ma_multi_malloc(0,
+		&new_entry, sizeof(struct Redirection_Entry),
+		&_key, key_len + 1,
+		&_host, host_len + 1,
+		&_user, user_len + 1,
+		NULL))
+	{
+		return;
+	}
+
+	new_entry->key = _key;
+	new_entry->host = _host;
+	new_entry->user = _user;
+	new_entry->next = NULL;
+	new_entry->prev = NULL;
+
+	strcpy(new_entry->key, key);
+	strcpy(new_entry->host, host);
+	strcpy(new_entry->user, user);
+	new_entry->port = port;
+
+	mutex_lock(redirect_mutex);
+
+	if (NULL == redirection_cache_root.next)
+	{
+		redirection_cache_root.next = new_entry;
+		new_entry->prev = &redirection_cache_root;
+		redirection_cache_rear = new_entry;
+	}
+	else
+	{
+		new_entry->next = redirection_cache_root.next;
+		new_entry->prev = &redirection_cache_root;
+
+		redirection_cache_root.next->prev = new_entry;
+		redirection_cache_root.next = new_entry;
+	}
+
+	redirection_cache_num++;
+
+	if (redirection_cache_num > MAX_CACHED_REDIRECTION_ENTRYS)
+	{
+		popingEntry = redirection_cache_rear;
+		redirection_cache_rear = redirection_cache_rear->prev;
+		redirection_cache_rear->next = NULL;
+
+		redirection_cache_num--;
+	}
+
+	mutex_unlock(redirect_mutex);
+
+	if (popingEntry)
+		free(popingEntry);
+
+	return;
+}
+
+void delete_redirection_entry(char* key)
+{
+	struct Redirection_Entry* del_entry = NULL;
+	struct Redirection_Entry* cur_entry = NULL;
+
+	mutex_lock(redirect_mutex);
+	cur_entry = redirection_cache_root.next;
+	while (cur_entry)
+	{
+		if (cur_entry->key)
+		{
+			if (!strcmp(cur_entry->key, key))
+			{
+				del_entry = cur_entry;
+				break;
+			}
+			cur_entry = cur_entry->next;
+		}
+		else break;
+	}
+
+	if (del_entry)
+	{
+		if (redirection_cache_rear == del_entry)
+		{
+			redirection_cache_rear = del_entry->prev;
+		}
+		del_entry->prev->next = del_entry->next;
+
+		if (del_entry->next)
+		{
+			del_entry->next->prev = del_entry->prev;
+		}
+
+		redirection_cache_num--;
+	}
+
+	mutex_unlock(redirect_mutex);
+
+	if (del_entry)
+		free(del_entry);
+}
+
+void* ma_multi_malloc(myf myFlags, ...)
```

Comment:
Better use an #ifdef, otherwise plugin_type=STATIC will fail (ma_multi_malloc is in ma_malloc.c)